### PR TITLE
feat: Add --location=<href> and globalThis.location

### LINKED ISSUE #4981: discussion: --location=<URL> option
In https://github.com/denoland/deno/issues/1657#issuecomment-567075831 an `--origin=https://example.com` option is proposed to select which `localStorage` isolate is used by the process. This could also be used for lots of other purposes.

For a consistent analogy we should extend this to a `--location=https://example.com/<path>` which selects the location for all web API purposes, including but not limited to:
- `window.location`.
- `fetch()` relative resolution root.
- `Worker` module relative resolution root.
- `localStorage` origin.
- `Origin` header for any built-in HTTP requests.
- Base URL for inline import maps.

Without an explicit `--location`, `globalThis.location` should throw, `localStorage` should be non-persistent and document-relative URLs for the other purposes should fail. This makes `--location` act as the gate requested for persistent storage: https://github.com/denoland/deno/issues/1657#issuecomment-612484399.

Not having a default might be controversial, but this lets us abstain from any opinionated design decisions for the first version.

---

JS API:

```js
Object.defineProperty(globalThis, "location", {
  get() {
    if (locationFromFlag == null) {
      throw new ReferenceError(
        `Access to "globalThis.location", run again with --location=<href>.`,
      );
    }
    return locationFromFlag;
  },
  // A similar setter would be applied to fields of `locationFromFlag`.
  set() {
    throw new TypeError(`Cannot set "globalThis.location".`);
  },
});
```

### PR DESCRIPTION
**Pull request by nayeemrmn**
Closes #4981.

https://github.com/nayeemrmn/deno/blob/location/docs/runtime/location_api.md

### CODE REVIEW

File: op_crates/web/lib.rs
Comment (nayeemrmn) [inline op_crates/web/lib.rs:21]: @bartlomieju There's a bug on Windows where any error with an `op_crates/web` source file in the stack will [panic on source mapping](https://github.com/denoland/deno/runs/1078336836).
```
> deno.exe eval "new Event() // Missing arg."
thread 'main' panicked at 'not implemented: Don't know how to create cache name for scheme: d', cli\disk_cache.rs:107:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
The specifier you pass here will be parsed as a URL during source mapping, which would cause an invalid scheme error for drive letter paths.

I've fixed that with `Url::from_file_path()` which correctly converts a drive letter file URL.

Comment (caspervonb) [reply op_crates/web/lib.rs:21]: Noticed this in coverage too, op_crates/web are the only files that have non url based specifiers when asking the inspector about things.

Comment (nayeemrmn) [reply op_crates/web/lib.rs:21]: The change I'm highlighting fixes it.

Comment (bartlomieju) [reply op_crates/web/lib.rs:21]: @nayeemrmn nice catch! Could you factor out this change to a separate PR so it could be landed independently? I'm not sure we'll go along with `--location` for 1.4.0, but looks like it fixes actual bug.

**Comment by caspervonb** [comment]
Raised this before with Nayeem in chat but API Semantics are fine I think, but we might have to allow for this to change the origin of contexts, so if we can better to see this behind --unstable so that we aren't immediately locked into the flags semantics.

This came up recently with an awkward hack for <unknown> origins, I think we've barely pried the lid off this can of worms.

**Comment by nayeemrmn** [comment]
We need the flag either way so it can run web scripts out of the box. A similar thing came up with service workers: https://discord.com/channels/684898665143206084/684898665151594506/741039703943413873. This is the main priority for `--location`.

As for whether or not we want to allow changing the location at runtime down the line... I'd want to hear the use cases. I suspect they aren't isomorphic with the web.

**Comment by caspervonb** [comment]
> We need the flag either way so it can run web scripts out of the box.

Yep thats fine, i'd just like to see the flag be unstable at first as all other new flags have been so we're not permanently locked into semantics we're just starting to unravel.

**Comment by kitsonk** [comment]
Sorry I haven't paid as much attention to this as I should.  A couple questions:

- Can the URI be a local file (`file:///path/to/file.html`)
- All of your examples in the documentation are just using just a hostname, with no path...  I assume something like `https://example.com/a/path/index.html` is acceptable?

**Comment by kitsonk** [comment]
Another thought, especially for workers, it would be really handy to have a shortcut to set it to the cwd.

**Comment by lucacasonato** [comment]
> Can the URI be a local file (file:///path/to/file.html)

This is probably the biggest question yet. One main use cases for having a window location, is being able to determine the script origin. Getting an origin for a file URL is explicitly not specified: https://url.spec.whatwg.org/#origin. Currently we do the same as Chrome, which is give all file URLs the same origin: `new URL("file:///a/b/c").origin == "file://"`. This sucks for web storage apis like localstorage, because all local scripts would share a storage bucket. Because of this I think we should disallow `file://` URLs in `--location`, at least for the first pass.

> I assume something like https://example.com/a/path/index.html is acceptable?

Yeah, although `https://example.com/a/path/index.html` and `https://example.com/b/foo/index.html` would share the origin `https://example.com`, and would this share a storage bucket.

**Comment by nayeemrmn** [comment]
> Can the URI be a local file (`file:///path/to/file.html`)

Yes, since you can get such a `location` by opening a file URL in a browser. It might be better to unsupport them at the level of origin-concerned consumers like `localStorage`. But I also don't mind removing this.

> Another thought, especially for workers, it would be really handy to have a shortcut to set it to the cwd.

Like allowing relative paths? As explained in the last section of the documentation, I think this only makes bad usage easier. People already have a way of specifying relatively located worker URLs with `import.meta.url`. It's part of the reason I didn't want a default.

**Comment by kitsonk** [comment]
> But I also don't mind removing this.

Actually I think it is handy, and local storage should do whatever local storage does when the origin is a local file.

**Comment by lucacasonato** [comment]
> local storage should do whatever local storage does when the origin is a local file

The `origin` of a local file is `file://`. That means all local files share a local storage bucket. That seems very, very unintuitive.

**Comment by kitsonk** [comment]
> That means all local files share a local storage bucket. That seems very, very unintuitive.

Lots of web standards are unintuitive... 😄   But that is something to debate on a local storage PR. 😉

**Comment by lucacasonato** [comment]
> But that is something to debate on a local storage PR

I disagree - this is not something we can push to the future, as this problem is not local storage specific. The core problem here is determining the window origin from the window location. We have to be aware of the consequences of allowing file URLs that when landing this PR. There is nothing we can do about it later in the local storage PR (or any other PR that uses window origin).

**Comment by kitsonk** [comment]
> I disagree - this is not something we can push to the future, as this problem is not local storage specific. The core problem here is determining the window origin from the window location. We have to be aware of the consequences of allowing file URLs as a window location when landing this PR. There is nothing we can do about it later in the local storage PR (or any other PR that uses window origin).

But I am not sure what you are arguing or proposing.  You can't magic up another origin out of thin air.  While I agree that if you set location, that you share the same storage bucket for all code that shares the same origin might be surprising for some, it is logical and I don't see any other alternative.  That is something to make clear in the documentation versus holding up what is consistent logic.  The alternative is to disallow `file:///` locations which seems illogical.

What we haven't addressed though would be how `file://some-host/` would be handled versus `file:///` URIs.  That would give different origins, but are confusing for things like `fetch`, because "where" is that file?

**Comment by lucacasonato** [comment]
> all code that shares the same origin might be surprising for some

That isn't my point - that is indeed totally expected. The unexpected part is that all file URLs share an origin. Most people likely expect that `file:///home/lucacasonato/myproject/index.html` and `file:///home/lucacasonato/secretproject/index.html` don't share a local storage bucket.

Also `file://` URLs would cause some other issues. If relative URLs in `fetch` based on the window location are possible, how would this interact with non http urls? We can't fetch `file://` URLs at the moment.

My main issue is that `file://` origins bring with it ton of edge cases. They are pretty much completely not standardised (see https://url.spec.whatwg.org/#origin) and their implementation and permissions vary greatly depending on user agent. I think we can figure this out in the long run, but for the short term it adds a lot of complexity for arguably very little value. I would prefer we do the initial implementation without support for `file://` window locations, and then slowly tackle this after we figured out how this feature is used in the wild.

**Comment by kitsonk** [comment]
Fair enough.  My thought was the benefit to workers, but as @nayeemrmn points out, there are better ways.  There is no harm in disallowing them for now, I agree.

**Comment by bartlomieju** [comment]
Core team discussed `--location` flag on several occasions, but we were unable to reach a consensus. 

Given that release is scheduled for Tuesday I think it's best solution to bump this PR to 1.6.0 (yes, I know it was already bumped from 1.4.0). 

There is no immediate need for this flag in upcoming release and the first feature that requires it is the `localStorage` implementation which is not gonna make it for 1.5.0 anyway.

File: cli/flags.rs
Comment (bartlomieju) [inline cli/flags.rs:336]: What's the reason for removing this function?

File: cli/flags.rs
Comment (bartlomieju) [inline cli/flags.rs:1534]: Can this function be turned into `clap` validator function in `Arg` declaration?

File: cli/standalone.rs
Comment (bartlomieju) [inline cli/standalone.rs:186]: If `location` is taken from `flags.location` below why is it not passed here?

File: cli/tests/unit/fetch_test.ts
Comment (bartlomieju) [inline cli/tests/unit/fetch_test.ts:233]: We should still run this test when `--location` is not present. Might have to add `location` option to `unitTest` definition 🤔  could you turn it into integration test in the mean time?

File: docs/runtime/workers.md
Comment (bartlomieju) [inline docs/runtime/workers.md:15]: This is true only for workers create from main worker. If I understand correctly all workers have location set to the resolved specifier, so for "2nd generation" workers one can always use relative module specifier.

File: cli/flags.rs
Comment (nayeemrmn) [reply cli/flags.rs:336]: The exiting one is removed because I wanted to unify that with the other errors in `main.rs`, under `unwrap_or_exit()`.

File: cli/standalone.rs
Comment (nayeemrmn) [reply cli/standalone.rs:186]: Ah, this is from before runtime flags were passable.

File: cli/flags.rs
Comment (bartlomieju) [reply cli/flags.rs:336]: I see, given number of lines changed this change caused I don't think it's worth it and using validator (https://github.com/denoland/deno/pull/7369#discussion_r552255305) would help as well

Comment (nayeemrmn) [reply cli/flags.rs:336]: No the line changes are from renaming `flags_from_vec_safe()` to the now available `flags_from_vec()`. I think it's better to only have one of these either way.

File: docs/runtime/workers.md
Comment (lucacasonato) [reply docs/runtime/workers.md]: I think 2nd gen workers still should be relative to top level window.location right? I'd have to check the spec.

**Review by ry — APPROVED** [review]
LGTM - sorry for the very long delay @nayeemrmn and thank you for seeing this through.

@lucacasonato reports this is needed for fetch WPT tests, hence the sudden movement on this.

File: cli/flags.rs
Comment (lucacasonato) [inline cli/flags.rs:1524]: I think the most recent consensus was that we shouldn't allow `file://` for the first pass, because it does not work with relative fetch. Can you remove this protocol from the supported list for now?

File: runtime/js/99_main.js
Comment (lucacasonato) [inline runtime/js/99_main.js:370]: On line 305 this is wrapped in a conditional. Why not here?

**Review by lucacasonato — CHANGES_REQUESTED** [review]

File: runtime/js/99_main.js
Comment (nayeemrmn) [reply runtime/js/99_main.js:370]: Workers are unconditionally given a location.

Comment (lucacasonato) [reply runtime/js/99_main.js:370]: What location? So we align with browsers here?

Comment (nayeemrmn) [reply runtime/js/99_main.js:370]: Yes, the location in a worker is the URL of its entrypoint module.

File: runtime/js/11_workers.js
Comment (lucacasonato) [inline runtime/js/11_workers.js:182]: `./foo.ts`, `/foo.ts`, and `../foo.ts` should all be relative to `baseUrl`. You should be able to remove this conditional.

**Review by lucacasonato — COMMENTED** [review]
One final comment.

File: runtime/js/11_workers.js
Comment (nayeemrmn) [reply runtime/js/11_workers.js:182]: Ah yeah, I need to catch all three of those. The conditional needs to stay though, the point is to let invalid bare specifiers (unless in classic mode) pass this and hit an error in Rust, rather than being based into a valid absolute specifier.

**Review by lucacasonato — APPROVED** [review]
LGTM. Thanks a bunch @nayeemrmn. Sorry this took so long!

**Pull request merged**
